## Part 3: Creating a customised chatbot

Based on part 1 and 2 in the earlier notebooks, I have extracted the necessary data that I need as context for building a customised chatbot. 
The following portion of this codebook will include:
- Part 3a: Creating a default chatbot without introducing any system prompt
- Part 3b: Creating an improved chatbot with the use of system prompt
- Conclusion

In [73]:
import os
import openai

from llama_index import Document, GPTVectorStoreIndex, ServiceContext
from llama_index.readers import SimpleDirectoryReader
from llama_index.llms import OpenAI
from llama_index.evaluation import DatasetGenerator

In [63]:
# Here we will need to use our own OpenAI API key. This key is removed due to privacy issue.
os.environ['OPENAI_API_KEY'] = "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"
openai.api_key = os.getenv("OPENAI_API_KEY")

In [74]:
data_dir="../extra_data"

In [76]:
filename_fn = lambda filename: {'file_name': filename}
my_docs = SimpleDirectoryReader(input_dir="../data", exclude_hidden=True, file_metadata=filename_fn).load_data()

print(f"Loaded {len(my_docs)} docs")

Loaded 125 docs


### Part 3a: Creating a default chatbot without any system prompt

In [30]:
# This is the original without any prompts for the chatbot
original_service_context = ServiceContext.from_defaults(
    llm=OpenAI(model="gpt-3.5-turbo", temperature=0))

In [31]:
original_index = GPTVectorStoreIndex.from_documents(documents=my_docs, service_context=original_service_context)

In [32]:
original_query_engine = original_index.as_query_engine()

#### Testing out some questions with the original query engine:

In [145]:
import time
start = time.time()
response = original_query_engine.query("How much salary must a candidate earn to be eligible for employment pass?")
print(response)
end = time.time()
print("")
print(f"This query took: {end-start} secs.")

A candidate must earn a fixed monthly salary starting from $5,000 to be eligible for an employment pass. However, candidates in the financial services sector need to earn higher salaries to qualify.

This query took: 8.021786212921143 secs.


In [146]:
start = time.time()
response = original_query_engine.query("How do I earn 20 points under the salary criteria?")
print(response)
end = time.time()
print("")
print(f"This query took: {end-start} secs.")

You can earn 20 points under the salary criteria by having a fixed monthly salary that is at or above the 90th percentile compared to the salary benchmarks by sector.

This query took: 8.367677211761475 secs.


In [147]:
start = time.time()
response = original_query_engine.query("Where can I find information about my company's diversity?")
print(response)
end = time.time()
print("")
print(f"This query took: {end-start} secs.")

You can find information about your company's diversity in the "Diversity" tab of the Workforce Insights tool, which can be accessed via the myMOM Portal.

This query took: 8.531136989593506 secs.


In [38]:
response = original_query_engine.query("I have a candidate from Fudan University, will she earn 20 points under the qualification criterion?")
print(response)

Yes, the candidate from Fudan University will earn 20 points under the qualification criterion.


In [71]:
response = original_query_engine.query("How much should my candidate aged 40 be earning from Construction sector to be awarded with 20 points?")
print(response)

Your candidate aged 40 should be earning $12,213 from the Construction sector in order to be awarded with 20 points.


In [89]:
response = original_query_engine.query("How much should my candidate from Accommodation be earning to be awarded 20 points?")
print(response)

Your candidate from the Accommodation sector should be earning $9,369 to be awarded 20 points.


**Based on the above queries that we have tested, the chatbot is able to come up with succint answers from the queries we have asked about qualifications, salary. But, we would ideally hope that the chatbot is able to include more context to how it derived the answer.** 

**As for the runtime, each query took about 8 to 9 secs to complete.** 

#### Generating the questions to measure the performance of the original query engine

In [149]:
# Shuffle the documents
import random

random.seed(42)
random.shuffle(my_docs)

In [150]:
question_gen_query = (
    "You are working at Ministry of Manpower focusing on eligibiity requirements of the employment pass. \
    There is a new complementarity assessment framework that will assess the eligibility of all prospective employment pass holders. \
    Your task is to setup all possible questions and requests, \
    using the provided context from documents on eligibility of employment pass, \
    formulate questions that capture important facts from the context. \
    Restrict the question to the context information provided."
)

In [152]:
dataset_generator = DatasetGenerator.from_documents(
    my_docs,
    question_gen_query=question_gen_query,
    service_context=original_service_context,
)

In [153]:
questions = dataset_generator.generate_questions_from_nodes(num=30)
print("Generated ", len(questions), " questions")

Generated  30  questions


In [154]:
with open("../qns_and_eval/original_evaluation_questions.txt", "w") as f:
    for question in questions:
        f.write(question + "\n")

In [155]:
original_contexts = []
original_answers = []

for question in questions:
    response = original_query_engine.query(question)
    original_contexts.append([x.node.get_content() for x in response.source_nodes])
    original_answers.append(str(response))

In [157]:
from datasets import Dataset
from ragas import evaluate
from ragas.metrics import answer_relevancy, faithfulness

original_ds = Dataset.from_dict(
    {
        "question": questions,
        "answer": original_answers,
        "contexts": original_contexts,
    }
)

original_result = evaluate(original_ds, [answer_relevancy, faithfulness])
print(original_result)

evaluating with [answer_relevancy]


100%|██████████████████████████████████████████████| 2/2 [04:42<00:00, 141.28s/it]


evaluating with [faithfulness]


  0%|                                                       | 0/2 [00:00<?, ?it/s]Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised Timeout: Request timed out: HTTPSConnectionPool(host='api.openai.com', port=443): Read timed out. (read timeout=600).
Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised Timeout: Request timed out: HTTPSConnectionPool(host='api.openai.com', port=443): Read timed out. (read timeout=600).
100%|█████████████████████████████████████████████| 2/2 [34:36<00:00, 1038.20s/it]


{'ragas_score': 0.8767, 'answer_relevancy': 0.9227, 'faithfulness': 0.8350}


In [53]:
questions = dataset_generator.generate_questions_from_nodes(num=50)
print("Generated ", len(questions), " questions")

Generated  50  questions


In [54]:
with open("../qns_and_eval/train_questions.txt", "w") as f:
    for question in questions:
        f.write(question + "\n")

In [55]:
eval_dataset_generator = DatasetGenerator.from_documents(
    my_docs[90:],  # In our training set we loaded 34 docs, so we will now use the remaining starting from 35
    question_gen_query=question_gen_query,
    service_context=original_service_context,
)

In [56]:
eval_questions = eval_dataset_generator.generate_questions_from_nodes(num=50)
print("Generated ", len(eval_questions), " questions")

Generated  50  questions


In [57]:
with open("../qns_and_eval/eval_questions.txt", "w") as f:
    for question in eval_questions:
        f.write(question + "\n")

#### Generating scores for training and evaluation based on original query engine:

In [58]:
train_contexts = []
train_answers = []

for question in questions:
    response = original_query_engine.query(question)
    train_contexts.append([x.node.get_content() for x in response.source_nodes])
    train_answers.append(str(response))

In [59]:
from datasets import Dataset
from ragas import evaluate
from ragas.metrics import answer_relevancy, faithfulness

original_ds = Dataset.from_dict(
    {
        "question": questions,
        "answer": train_answers,
        "contexts": train_contexts,
    }
)

train_result = evaluate(original_ds, [answer_relevancy, faithfulness])
print(train_result)

evaluating with [answer_relevancy]


100%|██████████████████████████████████████████████| 4/4 [13:26<00:00, 201.62s/it]


evaluating with [faithfulness]


100%|██████████████████████████████████████████████| 4/4 [32:16<00:00, 484.09s/it]


{'ragas_score': 0.8160, 'answer_relevancy': 0.9140, 'faithfulness': 0.7370}


In [60]:
eval_contexts = []
eval_answers = []

for eval_question in eval_questions:
    eval_response = original_query_engine.query(eval_question)
    eval_contexts.append([x.node.get_content() for x in eval_response.source_nodes])
    eval_answers.append(str(eval_response))

In [61]:
original_eval_ds = Dataset.from_dict(
    {
        "question": eval_questions,
        "answer": eval_answers,
        "contexts": eval_contexts,
    }
)

eval_result = evaluate(original_eval_ds, [answer_relevancy, faithfulness])
print(eval_result)

evaluating with [answer_relevancy]


100%|███████████████████████████████████████████████| 4/4 [04:13<00:00, 63.28s/it]


evaluating with [faithfulness]


100%|██████████████████████████████████████████████| 4/4 [24:01<00:00, 360.46s/it]


{'ragas_score': 0.8285, 'answer_relevancy': 0.9613, 'faithfulness': 0.7280}


### Part 3b: Improved chatbot with system prompt

In [91]:
# This is the improved service context with context_window and system prompt added for the chatbot
improved_service_context = ServiceContext.from_defaults(
    llm=OpenAI(model="gpt-3.5-turbo", temperature=0), 
    context_window=2048, 
    system_prompt = "You are an expert who understands the eligibility criteria of employment pass and your job is to answer questions related to the COMPASS and all relevant requirements. Keep your answers factual and provide more context. When asked about salary criteria or C1, include both the age and sector assumed if not provided before answering.")

In [92]:
improved_index = GPTVectorStoreIndex.from_documents(documents=my_docs, service_context=improved_service_context)

In [83]:
improved_query_engine = improved_index.as_query_engine()

#### Testing out some questions with the improved query engine:

In [165]:
import time
start = time.time()
response = improved_query_engine.query("How much salary must a candidate earn to be eligible for employment pass?")
print(response)
end = time.time()
print("")
print(f"This query took: {end-start} secs.")

A candidate must have a fixed monthly salary starting from $5,000 to be eligible for an Employment Pass. The salary increases progressively with age, up to $10,500 for those in the mid-40s. However, candidates in the financial services sector must earn at least $5,500, with the salary also increasing progressively with age up to $11,500 for those in the mid-40s.

This query took: 18.43025517463684 secs.


In [167]:
start = time.time()
response = improved_query_engine.query("How do I earn 20 points under the salary criteria?")
print(response)
end = time.time()
print("")
print(f"This query took: {end-start} secs.")

To earn 20 points under the salary criteria, your candidate's fixed monthly salary should be at or above the 90th percentile of the salary benchmarks by sector.

This query took: 8.806796073913574 secs.


In [168]:
start = time.time()
response = improved_query_engine.query("Where can I find information about my company's diversity?")
print(response)
end = time.time()
print("")
print(f"This query took: {end-start} secs.")

You can find information about your company's diversity in the "Diversity" tab of the Workforce Insights tool on the myMOM Portal. This tab shows you the top nationalities of PMETs (Professionals, Managers, Executives, and Technicians) in your firm, allowing you to assess the diversity of your workforce.

This query took: 6.932214736938477 secs.


In [90]:
start = time.time()
response = improved_query_engine.query("How much should my candidate be earning from Construction sector to be awarded 10 points?")
print(response)
end = time.time()
print("")
print(f"This query took: {end-start} secs.")

To be awarded 10 points in the Construction sector, your candidate should be earning a salary of $4,770 or above.


In [84]:
start = time.time()
response = improved_query_engine.query("How much should my candidate aged 40 be earning from Construction sector to be awarded with 20 points?")
print(response)
end = time.time()
print("")
print(f"This query took: {end-start} secs.")

Your candidate, aged 40, should be earning a minimum salary of $12,213 from the Construction sector in order to be awarded with 20 points.


In [87]:
start = time.time()
response = improved_query_engine.query("How much should my candidate be earning to be awarded 10 points?")
print(response)

Based on the provided context, the required salary for a candidate to be awarded 10 points varies depending on the age and sector. Could you please provide the age and sector of your candidate so that I can give you the specific salary requirement?


In [88]:
response = improved_query_engine.query("How much should my candidate from Accommodation be earning to be awarded 20 points?")
print(response)

Based on the provided context information for the Accommodation sector, the required salary for 20 points (90th percentile of local PMETs) varies depending on the age of the candidate. 

For candidates aged 37, the required salary for 20 points is $9,104. 
For candidates aged 38, the required salary for 20 points is $9,369. 
For candidates aged 39, the required salary for 20 points is $9,635. 
For candidates aged 40, the required salary for 20 points is $9,900. 
For candidates aged 41, the required salary for 20 points is $10,165. 
For candidates aged 42, the required salary for 20 points is $10,431. 
For candidates aged 43, the required salary for 20 points is $10,696. 
For candidates aged 44, the required salary for 20 points is $10,962. 
For candidates aged 45 and above, the required salary for 20 points is $11,227. 

Please note that these figures are specific to the Accommodation sector and are based on the 90th percentile of local PMETs.


In [109]:
response = improved_query_engine.query("How much should my candidate from Manufacturing be earning to be awarded 20 points?")
print(response)

Based on the provided context information for the Manufacturing sector, the required salary for a candidate to be awarded 20 points is as follows:

- For candidates aged 40: $13,811
- For candidates aged 41: $14,211
- For candidates aged 42: $14,611
- For candidates aged 43: $15,011
- For candidates aged 44: $15,411
- For candidates aged 45 and above: $15,811

Please note that these salary figures represent the 90th percentile of local PMETs (Professionals, Managers, Executives, and Technicians) in the Manufacturing sector.


#### Generating the training and evaluation questions

In [158]:
# Shuffle the documents
import random

random.seed(42)
random.shuffle(my_docs)

In [159]:
question_gen_query = (
    "You are working at Ministry of Manpower focusing on eligibiity requirements of the employment pass. \
    There is a new complementarity assessment framework that will assess the eligibility of all prospective employment pass holders. \
    Your task is to setup all possible questions and requests, \
    using the provided context from documents on eligibility of employment pass, \
    formulate questions that capture important facts from the context. \
    Restrict the question to the context information provided."
)

In [160]:
improved_dataset_generator = DatasetGenerator.from_documents(
    my_docs,
    question_gen_query=question_gen_query,
    service_context=improved_service_context,
)

In [161]:
improved_questions = improved_dataset_generator.generate_questions_from_nodes(num=30)
print("Generated ", len(questions), " questions")

Generated  30  questions


In [162]:
with open("../qns_and_eval/improved_evaluation_questions.txt", "w") as f:
    for question in improved_questions:
        f.write(question + "\n")

In [163]:
improved_eval_contexts = []
improved_eval_answers = []

for question in improved_questions:
    improved_eval_response = improved_query_engine.query(question)
    improved_eval_contexts.append([x.node.get_content() for x in improved_eval_response.source_nodes])
    improved_eval_answers.append(str(improved_eval_response))

In [164]:
from datasets import Dataset
from ragas import evaluate
from ragas.metrics import answer_relevancy, faithfulness

improved_eval_ds = Dataset.from_dict(
    {
        "question": improved_questions,
        "answer": improved_eval_answers,
        "contexts": improved_eval_contexts,
    }
)

improved_eval_result = evaluate(improved_eval_ds, [answer_relevancy, faithfulness])
print(improved_eval_result) 

evaluating with [answer_relevancy]


100%|███████████████████████████████████████████████| 2/2 [03:05<00:00, 92.52s/it]


evaluating with [faithfulness]


  0%|                                                       | 0/2 [00:00<?, ?it/s]Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised Timeout: Request timed out: HTTPSConnectionPool(host='api.openai.com', port=443): Read timed out. (read timeout=600).
Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised Timeout: Request timed out: HTTPSConnectionPool(host='api.openai.com', port=443): Read timed out. (read timeout=600).
100%|█████████████████████████████████████████████| 2/2 [48:17<00:00, 1448.83s/it]


{'ragas_score': 0.8310, 'answer_relevancy': 0.9606, 'faithfulness': 0.7322}


In [96]:
improved_questions = improved_dataset_generator.generate_questions_from_nodes(num=50)
print("Generated ", len(questions), " questions")

Generated  50  questions


In [97]:
with open("../qns_and_eval/improved_train_questions.txt", "w") as f:
    for question in improved_questions:
        f.write(question + "\n")

In [98]:
improved_eval_dataset_generator = DatasetGenerator.from_documents(
    my_docs[90:],  # In our training set we loaded 90 docs, so we will now use the remaining for eval.
    question_gen_query=question_gen_query,
    service_context=improved_service_context,
)

In [100]:
improved_eval_questions = improved_eval_dataset_generator.generate_questions_from_nodes(num=50)
print("Generated ", len(improved_eval_questions), " questions")

Generated  50  questions


In [101]:
with open("../qns_and_eval/improved_eval_questions.txt", "w") as f:
    for question in improved_eval_questions:
        f.write(question + "\n")

#### Generating scores for training and evaluation based on improved query engine:

In [102]:
improved_train_contexts = []
improved_train_answers = []

for question in improved_questions:
    improved_response = improved_query_engine.query(question)
    improved_train_contexts.append([x.node.get_content() for x in improved_response.source_nodes])
    improved_train_answers.append(str(improved_response))

In [103]:
from datasets import Dataset
from ragas import evaluate
from ragas.metrics import answer_relevancy, faithfulness

improved_ds = Dataset.from_dict(
    {
        "question": improved_questions,
        "answer": improved_train_answers,
        "contexts": improved_train_contexts,
    }
)

improved_train_result = evaluate(improved_ds, [answer_relevancy, faithfulness])
print(improved_train_result)

evaluating with [answer_relevancy]


 25%|███████████▌                                  | 1/4 [11:00<33:00, 660.25s/it]Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised ServiceUnavailableError: The server is overloaded or not ready yet..
Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised ServiceUnavailableError: The server is overloaded or not ready yet..
Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised Timeout: Request timed out: HTTPSConnectionPool(host='api.openai.com', port=443): Read timed out. (read timeout=600).
Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised ServiceUnavailableError: The server is overloaded or not ready yet..
 50%|███████████████████████                       | 2/4 [25:18<25:53

evaluating with [faithfulness]


  0%|                                                       | 0/4 [00:00<?, ?it/s]Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised APIError: Bad gateway. {"error":{"code":502,"message":"Bad gateway.","param":null,"type":"cf_bad_gateway"}} 502 {'error': {'code': 502, 'message': 'Bad gateway.', 'param': None, 'type': 'cf_bad_gateway'}} {'Date': 'Mon, 06 Nov 2023 13:27:26 GMT', 'Content-Type': 'application/json', 'Content-Length': '84', 'Connection': 'keep-alive', 'X-Frame-Options': 'SAMEORIGIN', 'Referrer-Policy': 'same-origin', 'Cache-Control': 'private, max-age=0, no-store, no-cache, must-revalidate, post-check=0, pre-check=0', 'Expires': 'Thu, 01 Jan 1970 00:00:01 GMT', 'Server': 'cloudflare', 'CF-RAY': '821d9de02bb84097-SIN', 'alt-svc': 'h3=":443"; ma=86400'}.
Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised Timeout: Reque

{'ragas_score': 0.8557, 'answer_relevancy': 0.9429, 'faithfulness': 0.7833}


In [104]:
improved_eval_contexts = []
improved_eval_answers = []

for improved_eval_question in improved_eval_questions:
    improved_eval_response = improved_query_engine.query(improved_eval_question)
    improved_eval_contexts.append([x.node.get_content() for x in improved_eval_response.source_nodes])
    improved_eval_answers.append(str(improved_eval_response))

In [114]:
improved_eval_ds = Dataset.from_dict(
    {
        "question": improved_eval_questions,
        "answer": improved_eval_answers,
        "contexts": improved_eval_contexts,
    }
)

improved_eval_result = evaluate(improved_eval_ds, [answer_relevancy, faithfulness])
print(improved_eval_result)

evaluating with [answer_relevancy]


100%|███████████████████████████████████████████████| 4/4 [05:00<00:00, 75.18s/it]


evaluating with [faithfulness]


 50%|███████████████████████                       | 2/4 [17:06<18:17, 548.57s/it]Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised Timeout: Request timed out: HTTPSConnectionPool(host='api.openai.com', port=443): Read timed out. (read timeout=600).
Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised APIError: Bad gateway. {"error":{"code":502,"message":"Bad gateway.","param":null,"type":"cf_bad_gateway"}} 502 {'error': {'code': 502, 'message': 'Bad gateway.', 'param': None, 'type': 'cf_bad_gateway'}} {'Date': 'Tue, 07 Nov 2023 13:44:42 GMT', 'Content-Type': 'application/json', 'Content-Length': '84', 'Connection': 'keep-alive', 'X-Frame-Options': 'SAMEORIGIN', 'Referrer-Policy': 'same-origin', 'Cache-Control': 'private, max-age=0, no-store, no-cache, must-revalidate, post-check=0, pre-check=0', 'Expires': 'Thu, 01 Jan 1970 00:0

{'ragas_score': 0.7730, 'answer_relevancy': 0.9860, 'faithfulness': 0.6357}


### Conclusion

Based on the execution time, the RAG score metrics as well as the quality of content provided can be summarised in the table below.

| Query engine         | RAGAS Score | Answer Relevancy | Faithfulness | 
|----------------------|-------------|------------------|--------------|
| Original (GPT-3.5-Turbo) | 0.8160 | 0.9140        | 0.7370       | 
| Improved with system prompt (GPT-3.5-Turbo) | 0.8557| 0.9429      | 0.7833 | 

In [107]:
#### We store the vectors of all the documents that we have curated for our chatbot that is to be deployed on streamlit
improved_index.storage_context.persist(persist_dir="../streamlit/improved_index.vecstore")